### PR TITLE
feat(ui):Admin users can do bulk delete component/releases

### PR DESCRIPTION
**Pull request by KoukiHama**
Signed-off-by: Kouki Hama <kouki1.hama@toshiba.co.jp>

[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? (*Refer to issue here*)
> * Did you add or update any new dependencies that are required for your change?

Issue:  #1649

Add a function to collectively delete releases connected by links starting from the specified release.

https://user-images.githubusercontent.com/39894802/193703151-ac027f89-b4fc-439e-9d25-8f9b672feb49.mp4

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

## How to test  
1. Preparing for the test  
    - Add the following options to sw360.properties before starting the system.  
        ```
        bulk.release.deleting.enabled=true  
        admin.private.project.access.enabled=true  
        ```
1. Executing Bulk delete from Release detail view  
    1. After starting the system, sign in as Admin user and register releases with the following structure.
        image
    1. Open the "DeleteLibrary004 1.00" release detail view and click the "Bulk Delete" button.  
    1. Click the "Bulk Delete Execution" button on Bulk delete preview screen.  
    1. Confirm that the following records are deleted.  
        - DeleteLibrary004 1.00  
        - DeleteLibrary005 1.00  
        - DeleteLibrary006 1.00  
    1. Confirm that the following records are not deleted.  
        - ExcludeLibrary001 1.00  
        - RemainLibrary001 1.00  

1. Executing Bulk delete from Component detail view  
    1. After starting the system, sign in as Admin user and register releases with the following structure.
        image
    1. Open the "DeleteLibrary004" component detail view and click the "Release Overview" menu.
    1. Click the "DeleteLibrary004 1.00" trash can button.  
    1. Click the "Bulk Delete Execution" button on Bulk delete preview screen.  
    1. Confirm that the following records are deleted.  
        - DeleteLibrary004 1.00  
        - DeleteLibrary005 1.00  
        - DeleteLibrary006 1.00  
    1. Confirm that the following records are not deleted.  
        - ExcludeLibrary001 1.00  
        - RemainLibrary001 1.00  

1. Operations by users without administrator privileges 
    1. After starting the system, sign in as a user without Admin privileges.
    1. Open a release detail view and confirm that the "Bulk Delete" button is not displayed.
    1. Confirm that Bulk delete function is not executed by the trash can button on a component detail view.  

# Overview of improvements

## Purpose  
- Add a function to collectively delete releases connected by links starting from the specified release.  
- Do not delete releases that are referenced by non-dependent records.
- Delete components related with releases deleted.  
- This function can only be performed by admin users.  
- This function can be turned on/off in "sw360.properties".  
    If you enable the function, you must also enable Admin private access function.  
    Options in "sw360.properties"  
    ```
    bulk.release.deleting.enabled=true  
    admin.private.project.access.enabled=true  
    ```

## The outline of changes

1. Add "Bulk Delete" button to Release detail view.  
   image
1. Change the behavior of trash can buttons in Release Overview menu on Component detail view.  
    If this function is enabled and you are Admin user, clicking the trash can button will trigger Bulk delete function for the release.  
    image
1. Add Bulk delete preview screen  
    When this function is executed, the following preview screen is displayed.  
    image
    
    On the preview screen, you can check the releases connected by the link starting from the specified release in the tree table. It also displays components related to releases in the tree.  

    "Operation" column indicates the operation when the process is executed.  
    |Operation|Description|
    |---|---|
    |Deleted|Deleted records|
    |Remained|Excluded records<BR>For example, records referenced by non-dependent records, components related to non-dependent releases, etc.|
    
    Click "Bulk Delete Execution" button to perform Bulk delete.  

1. Add Bulk delete result screen  
    When you execute Bulk delete, the following result screen will be displayed.  
   image

    On the result screen, as in the preview screen, you can check the releases connected by the link starting from the specified release in the tree table. It also displays components related to releases in the tree.   
    Successful records are given "-Deleted Item-" label.  

    "Result" column indicates execution results.  

    |Result|Description|
    |---|---|
    |Deleted|Deleted records|
    |Remained|Excluded records|
    |Conflicted|Records that conflicted with other users' requests during processing|
    |Error|Records that cannot be deleted for other reasons|

## Points of code change    

1. Backend  
    - Class diagram  
        Add BulkDeleteUtil class that contains the business logic of Bulk delete function to org.eclipse.sw360.datahandler.db package.  

        An instance of BulkDeleteUtil class is created when ComponentDatabaseHandler is initialized, and is retained as an instance member of ComponentDatabaseHandler until system termination.  

        BulkDeleteUtil handles bulk deletion requests received by ComponentDatabaseHandler.  

        image

    - Bulk delete flow  
        - The basic flow  
            The link structure is maintained as much as possible considering retry in case of failure due to an error.  
            Group the records that are the leaves of the link structure (red rectangles in the figure) and delete them all at once.  
           image
        - Cases where external referencing records exist  
            Records that are referenced by records outside the dependency relationship are excluded from the link structure.
           image
        - Cases where records cannot be deleted due to conflicts or errors  
            Records that cannot be deleted due to errors or revision update by other users are set to Error or Conflicted ​​status and excluded from deletion targets.  
            The link from the root record is left so that it can be retried after the error cause is resolved.  
          image
    - Bulk delete sequence  
        The main processing of Bulk delete function is BulkDeleteUtil.deleteBulkRelease method.  
        - Arguments
            |Name|Type|Description|
            |---|---|---|
            |releaseId|String|ID of the release that will be the starting point for Bulk delete|
            |user|User|User information |
            |isPreview|boolean|Flag to toggle Bulk delete or Bulk delete preview.<BR>If isPreview is true, do not update DB.|
            <BR>
        - Return value  
            Execution result is returned as BulkOperationNode.  
            BulkOperationNode has a nested structure by childList (list of BulkOperationNode), and expresses a network of releases connected by links.  
            <BR>
        - Sequence  
           image
            <BR>
1. Frontend  
    - The rendering process            
        The frontend expands BulkOperationNode returned by the backend server and displays linked releases in the tree table.
    - The rendering class  
        org.eclipse.sw360.portal.portlets.components.ComponentPortlet  

## Change of data structure    
Add a data class to represent a group of records connected by links starting from the specified record.

- BulkOperationNode  
    A data class for representing one record connected by links.
    It has the following attributes.

    |Name|Type|Description|
    |---|---|---|
    |id|String|ID of this record.|
    |name|String|The name of this record|
    |version|String|The version name of this record|
    |type|BulkOperationNodeType|The type of this record|
    |parentId|String|ID of the record that references this record|
    |childList|List\<BulkOperationNode\>|List of BulkOperationNode referenced from this record|
    |state|BulkOperationResultState|The result of executing bulk operation on this record|
    |additionalData|Map\<String, String\>|The map object that holds additional data for this record|

- BulkOperationNodeType  
    Enum for representing the type of BulkOperationNode.
    |Name|Value|
    |---|---|
    |PROJECT|0|
    |COMPONENT|1|
    |RELEASE|2|

- BulkOperationResultState  
    Enum for representing the result of bulk operations.
    |Name|Value|
    |---|---|
    |SUCCEEDED|0|
    |FAILED|1|
    |CONFLICTED|2|
    |EXCLUDED|3|

## Migration of DB  
There is no database schema change due to the addition of Bulk delete function.  

## limitations
Nothing special to mention.

### CODE REVIEW

**Comment by KoukiHama** [comment]
fixed conflict

**Comment by KoukiHama** [comment]
fixed conflict 
could you check it?
@ag4ums

**Comment by nikkuma7** [comment]
Hi, getting below error while click on bulk delete.
And also, this branch has conflict issue. 
image
image

**Comment by KoukiHama** [comment]
Thank you for checking.
I will check and fix them soon.

**Comment by nikkuma7** [comment]
@KoukiHama , Now bulk delete button not showing.
image

**Comment by KoukiHama** [comment]
sorry. i am still fixing. i will mention when i have done every work.

**Comment by KoukiHama** [comment]
I checked again. I was successful without issue, (from make making releases to deleting them)
Could you build again  [4e3dc06](https://github.com/eclipse-sw360/sw360/commit/4e3dc062b2dd4a66e651b973c84a92d7914926eb) and check if these options in sw360.property are true? @nikkuma7 

and if you find issue, could you tell me your logs, and when you find the issue in "How to test"?
 
bulk.release.deleting.enabled=true  
admin.private.project.access.enabled=true  

MicrosoftTeams-image (3)
MicrosoftTeams-image (5)
MicrosoftTeams-image (6)

**Comment by nikkuma7** [comment]
Test successful.
image
image

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/BulkDeleteUtil.java [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+    private AttachmentConnector attachmentConnector;
+    private AttachmentDatabaseHandler attachmentDatabaseHandler;
+    private DatabaseHandlerUtil dbHandlerUtil;    
+    
+    private BulkDeleteUtilInspector inspector;
+    
+    public interface BulkDeleteUtilInspector {
+        void checkVariables(Map<String, Release> releaseMap, Map<String, Component> componentMap,Map<String, Boolean> externalLinkMap, Map<String, List<String>> referencingReleaseIdsMap);
+        void checkLoopState(int loopCount, Map<String, Release> releaseMap, Map<String, Component> componentMap, Map<String, BulkOperationResultState> resultStateMap);
+        void checkleafReleaseIdsInLoop(int loopCount, Set<String> leafReleaseIds);
```
Comment (rudra-superrr) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/BulkDeleteUtil.java:74]: change checkleafReleaseIdsInLoop to checkLeafReleaseIdsInLoop

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/BulkDeleteUtil.java [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+        
+        //create a linked component map
+        Map<String, Component> allComponentMap = getComponenMap(allLinkedReleaseMap);
+        Map<String, Component> workComponentMap = getDuplicatedComponentMap(allComponentMap);
+        
+        if (inspector != null) inspector.checkVariables(workReleaseMap, workComponentMap, externalLinkMap, referencingReleaseIdsMap);
+        
+        //delete linked releases
+        boolean isCompleted = false;
+        for (int i=0; i<LOOP_MAX; i++) {
```
Comment (rudra-superrr) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/BulkDeleteUtil.java:148]: Use proper variable name for i.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/ErrorMessages.java [hunk 3]
```diff
@@ -68,6 +69,9 @@ public class ErrorMessages {
     public static final String OBLIGATION_NOT_UPDATED = "Obligation could not be updated.";
     public static final String VENDOR_DUPLICATE = "A vendor with the same name already exists.";
     public static final String ERROR_VENDOR = "Error: Invalid vendor Name or Url.";
+    public static final String ERROR_BULK_DELITING = "Error while bulk deleting";
+    public static final String ERROR_BULK_DELITING_IN_BACKEND = "Error while bulk deleting in backend.";
```
Comment (rudra-superrr) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/ErrorMessages.java:73]: Make it ERROR_BULK_DELETING and ERROR_BULK_DELETING_IN_BACKEND

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkRelease.jsp [hunk 4]
```diff
@@ -0,0 +1,122 @@
+<%--
+  ~ Copyright Siemens AG, 2013-2017, 2019. Part of the SW360 Portal Project.
+  ~ With modifications by Bosch Software Innovations GmbH, 2016.
+  ~
+  ~ This program and the accompanying materials are made
+  ~ available under the terms of the Eclipse Public License 2.0
+  ~ which is available at https://www.eclipse.org/legal/epl-2.0/
+  ~
+  ~ SPDX-License-Identifier: EPL-2.0
+--%>
```
Comment (rudra-superrr) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkRelease.jsp:10]: Update the license header

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkReleasePreview.jsp [hunk 5]
```diff
@@ -0,0 +1,118 @@
+<%--
+  ~ Copyright Siemens AG, 2013-2017, 2019. Part of the SW360 Portal Project.
+  ~ With modifications by Bosch Software Innovations GmbH, 2016.
+  ~
+  ~ This program and the accompanying materials are made
+  ~ available under the terms of the Eclipse Public License 2.0
+  ~ which is available at https://www.eclipse.org/legal/epl-2.0/
+  ~
+  ~ SPDX-License-Identifier: EPL-2.0
+--%>
```
Comment (rudra-superrr) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkReleasePreview.jsp:10]: Update the license header.

Comment (KoukiHama) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkReleasePreview.jsp:9]: done

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkRelease.jsp (hunk 4 above)
Comment (KoukiHama) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkRelease.jsp:9]: done

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/ErrorMessages.java (hunk 3 above)
Comment (KoukiHama) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/ErrorMessages.java]: done

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/BulkDeleteUtil.java (hunk 2 above)
Comment (KoukiHama) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/BulkDeleteUtil.java]: Done. I use loopCount.
Note : In other SW360 codes, 'i' is often used. This is certainly not a good idea.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/BulkDeleteUtil.java (hunk 1 above)
Comment (KoukiHama) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/BulkDeleteUtil.java]: Done

**Comment by KoukiHama** [comment]
@rudra-superrr  Thank you. I solved  them

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkReleasePreview.jsp (hunk 5 above)
Comment (rudra-superrr) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkReleasePreview.jsp:9]: There are four newly generated files, can you make the license header info uniform for all four.

Comment (KoukiHama) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/components/deleteBulkReleasePreview.jsp:9]: Ah ,  OK I will set a uniform header for all 4 files.

**Comment by rudra-superrr** [comment]
Looks good.

**Pull request merged**